l10n_it_delivery_note/models/sale_order.py

```diff
@@ -0,0 +1,207 @@
+        order_lines = self.mapped("order_line").filtered(
+            lambda l: l.is_invoiced and l.delivery_note_line_ids
+        )
+
+        delivery_note_lines = order_lines.mapped("delivery_note_line_ids").filtered(
+            lambda l: l.is_invoiceable
+        )
+        delivery_notes = delivery_note_lines.mapped("delivery_note_id")
+
+        ready_delivery_notes = delivery_notes.filtered(
+            lambda n: n.state != DOMAIN_DELIVERY_NOTE_STATES[0]
+        )
+
+        draft_delivery_notes = delivery_notes - ready_delivery_notes
+        draft_delivery_note_lines = (
+            draft_delivery_notes.mapped("line_ids") & delivery_note_lines
+        )
+
+        ready_delivery_note_lines = delivery_note_lines - draft_delivery_note_lines
+
+        #
+        # TODO: È necessario gestire il caso di fatturazione splittata
+        #        di una stessa riga d'ordine associata ad una sola
+        #        picking (e di conseguenza, ad un solo DdT)?
+        #       Può essere, invece, un caso "borderline"
+        #        da lasciar gestire all'operatore?
+        #       Personalmente, non lo gestirei e delegherei
+        #        all'operatore questa responsabilità...
+        #
+
+        draft_delivery_note_lines.write(
+            {"invoice_status": DOMAIN_INVOICE_STATUSES[0], "sale_line_id": None}
+        )
+
+        ready_delivery_note_lines.write({"invoice_status": DOMAIN_INVOICE_STATUSES[2]})
+        ready_delivery_notes.write(
+            {"invoice_ids": [(4, invoice_id) for invoice_id in invoice_ids]}
+        )
+
+        ready_delivery_notes._compute_invoice_status()
+
+    def _generate_delivery_note_lines(self, invoice_ids):
+        invoices = self.env["account.move"].browse(invoice_ids)
+        invoices.update_delivery_note_lines()
+
+    def _create_invoices(self, grouped=False, final=False, date=None):
+        invoice_ids = super()._create_invoices(grouped=grouped, final=final, date=date)
+
+        self._assign_delivery_notes_invoices(invoice_ids.ids)
+        self._generate_delivery_note_lines(invoice_ids.ids)
+
+        return invoice_ids
+
+    def goto_delivery_notes(self, **kwargs):
+        delivery_notes = self.mapped("delivery_note_ids")
+        action = self.env.ref(
+            "l10n_it_delivery_note.stock_delivery_note_action"
+        ).read()[0]
+        action.update(kwargs)
+
+        if len(delivery_notes) > 1:
+            action["domain"] = [("id", "in", delivery_notes.ids)]
+
+        elif len(delivery_notes) == 1:
+            action["views"] = [
+                (
+                    self.env.ref(
+                        "l10n_it_delivery_note.stock_delivery_note_form_view"
+                    ).id,
+                    "form",
+                )
+            ]
+            action["res_id"] = delivery_notes.id
+
+        else:
+            action = {"type": "ir.actions.act_window_close"}
+
+        return action
+
+
+class SaleOrderLine(models.Model):
```

Comment:
Rimuovere perché è presente lo stesso codice nel file sale_order_line.py

Reply:
rimosso class SaleOrderLine in file sale_order.py